File: heap/1046.java
import java.util.PriorityQueue;

class Solution {
    public int lastStoneWeight(int[] stones) {
        PriorityQueue<Integer> pq = new PriorityQueue<>(new Comparator<Integer>(){
            public int compare(Integer i1, Integer i2) {
                return i2 - i1;
            }
        });
        for (int i : stones) {
            pq.offer(i);
        }
        while (pq.size() > 1) {
            int a = pq.poll();
            int b = pq.poll();
            if (a < b) {
                pq.offer(b-a);
            }else if (a > b) {
                pq.offer(a-b);
            }
        }
        if (pq.size() == 0) {
            return 0;
        }
        return pq.poll();
    }
}

File: heap/1046.cc
class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {
        auto cmp = [](int left, int right) {
            return left < right;
        };
        std::priority_queue<int, std::vector<int>, decltype(cmp)> q(cmp);
        for (const auto& s : stones) {
            q.emplace(s);
        }
        while (q.size() > 1) {
            int s1 = q.top();
            q.pop();
            int s2 = q.top();
            q.pop();
            if (s1 != s2) {
                q.emplace(std::abs(s1-s2));
            }
        }
        if (q.empty()) return 0;
        return q.top();
    }
};
